File: python/complex_text/complex_text/complex_text.py
# ============================================================================
#  complex_text -- Python (ctypes) port of
#  examples\Vb6\complex_text\complex_text\complex_text.bas
#  Complex text layout of a right-to-left (Pashto) text. The text is loaded as
#  UTF-16 and laid out with the wide version of WriteFTextEx().
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))
TEST = os.path.join(HERE, "..", "..", "..", "test_files")


@L.TErrorProc
def err_proc(data, errcode, errmsg, errtype):
    if errmsg:
        print(errmsg.decode("latin-1", "replace"))
    return 0  # try to continue on error


def get_file_buffer(filename):
    try:
        with open(filename, "rb") as fh:
            data = fh.read()
    except OSError:
        return ""
    return data.decode("utf-16-le", "replace")


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, err_proc)
    L.pdfCreateNewPDFA(pdf, b"")  # output opened later

    txt = get_file_buffer(os.path.join(TEST, "pashto.txt"))

    L.pdfSetPageCoords(pdf, L.pcTopDown)
    # Enable complex text layout
    L.pdfSetGStateFlags(pdf, L.gfComplexText, 0)
    L.pdfSetBidiMode(pdf, L.bmRightToLeft)

    L.pdfAppend(pdf)

    # The font must be loaded with cpUnicode.
    L.pdfSetFontA(pdf, b"Arial", L.fsRegular, 10.0, 1, L.cpUnicode)
    L.pdfSetLeading(pdf, L.pdfGetTypoLeading(pdf))
    L.pdfWriteFTextExW(pdf, 50.0, 50.0, L.pdfGetPageWidth(pdf) - 100.0,
                       L.pdfGetPageHeight(pdf) - 100.0, L.taJustify, txt)

    L.pdfEndPage(pdf)

    out_file = os.path.join(HERE, "out.pdf")
    if L.pdfHaveOpenDoc(pdf) != 0:
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
    if L.pdfCloseFile(pdf) != 0:
        print('PDF file "' + out_file + '" successfully created!')
    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: csharp/complex_text/complex_text/complex_text.cs
//  complex_text -- C# port of examples\Vb6\complex_text\complex_text\complex_text.bas
//  Complex text layout of a right-to-left (Pashto) text loaded as UTF-16 and
//  laid out with the wide version of WriteFTextEx().
using System;
using System.IO;
using System.Text;
using LumasPdfSdk;

class ComplexText
{
    static TErrorProc _err = ErrProc;
    static int ErrProc(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;
    }

    static string GetFileBuffer(string fileName)
    {
        try { return Encoding.Unicode.GetString(File.ReadAllBytes(fileName)); }
        catch { return ""; }
    }

    static void Main()
    {
        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfCreateNewPDFW(pdf, "");        // We create no PDF file in this example

        string txt = GetFileBuffer(Path.GetFullPath(Path.Combine(
            AppDomain.CurrentDomain.BaseDirectory, "../../../test_files/pashto.txt")));

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);
        LumasPdf.pdfSetGStateFlags(pdf, LumasPdfConsts.gfComplexText, false);
        LumasPdf.pdfSetBidiMode(pdf, (int)TPDFBidiMode.bmRightToLeft);

        LumasPdf.pdfAppend(pdf);
        LumasPdf.pdfSetFontW(pdf, "Arial", LumasPdfConsts.fsRegular, 10.0, true, TCodepage.cpUnicode);
        LumasPdf.pdfSetLeading(pdf, LumasPdf.pdfGetTypoLeading(pdf));
        LumasPdf.pdfWriteFTextExW(pdf, 50.0, 50.0,
            LumasPdf.pdfGetPageWidth(pdf) - 100.0, LumasPdf.pdfGetPageHeight(pdf) - 100.0,
            (int)LumasPdfConsts.taJustify, txt);
        LumasPdf.pdfEndPage(pdf);

        string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
        }
        if (LumasPdf.pdfCloseFile(pdf))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        LumasPdf.pdfDeletePDF(pdf);
    }
}
